src/librustc_privacy/lib.rs

```diff
@@ -974,7 +974,7 @@ impl<'a, 'tcx: 'a> TypeVisitor<'tcx> for SearchInterfaceForPrivateItemsVisitor<'
         if let ty::TyProjection(ref proj) = ty.sty {
             // Avoid calling `visit_trait_ref` below on the trait,
             // as we have already checked the trait itself above.
-            proj.trait_ref.super_visit_with(self)
+            proj.substs.visit_with(self)
```

Comment:
Oh so I just realized, this file is *the only remaining user* of `visit_trait_ref`, which means you can remove that method (making `TypeVisitor` minimal again), and instead move `visit_trait_ref` implementations to inherent methods that you call explicitly from where they would get invoked otherwise.

Oh and this special-case at the end is unnecessary now! `ty.super_visit_with(self)` does the right thing in all cases.